miden-lib/src/notes/mod.rs

```diff
@@ -4,96 +4,104 @@ use miden_objects::{
-            vec![target.into(), recall_height.into(), ZERO, ZERO],
-        ),
-        Script::SWAP { asset, serial_num } => {
-            let recipient = build_p2id_recipient(sender, serial_num)?;
-            let asset_word: Word = asset.into();
-            (
-                ProgramAst::from_bytes(swap_bytes).map_err(NoteError::NoteDeserializationError)?,
-                vec![
-                    recipient[0],
-                    recipient[1],
-                    recipient[2],
-                    recipient[3],
-                    asset_word[0],
-                    asset_word[1],
-                    asset_word[2],
-                    asset_word[3],
-                    sender.into(),
-                    ZERO,
-                    ZERO,
-                    ZERO,
-                ],
-            )
-        },
-    };
-
-    let (note_script, _) = NoteScript::new(note_script_ast, &note_assembler)?;
-
-    Note::new(note_script.clone(), &inputs, &assets, serial_num, sender, tag.unwrap_or(ZERO))
-}
+    let swap_script_ast =
+        ProgramAst::from_bytes(swap_bytes).map_err(NoteError::NoteDeserializationError)?;
 
-/// Utility function generating RECIPIENT for the P2ID note script created by the SWAP script
-fn build_p2id_recipient(target: AccountId, serial_num: Word) -> Result<Digest, NoteError> {
-    // TODO: add lazy_static initialization or compile-time optimization instead of re-generating
-    // the script hash every time we call the SWAP script
-    let assembler = TransactionKernel::assembler();
+    let requested_serial_num = rng.draw_word();
+
+    let recipient = utils::build_p2id_recipient(sender, requested_serial_num.clone())?;
+
+    let asset_word: Word = requested_asset.into();
 
-    let p2id_bytes = include_bytes!(concat!(env!("OUT_DIR"), "/assets/P2ID.masb"));
+    let swap_inputs = vec![
+        recipient[0],
+        recipient[1],
+        recipient[2],
+        recipient[3],
+        asset_word[0],
+        asset_word[1],
+        asset_word[2],
+        asset_word[3],
+        sender.into(),
+        ZERO,
+        ZERO,
+        ZERO,
+    ];
 
-    let note_script_ast =
-        ProgramAst::from_bytes(p2id_bytes).map_err(NoteError::NoteDeserializationError)?;
+    let (note_script, _) = NoteScript::new(swap_script_ast, &assembler)?;
 
-    let (note_script, _) = NoteScript::new(note_script_ast, &assembler)?;
+    let tag: Felt = Felt::new(0);
 
-    let script_hash = note_script.hash();
+    let serial_num = rng.draw_word();
 
-    let serial_num_hash = Hasher::merge(&[serial_num.into(), Digest::default()]);
+    let swap_note = Note::new(
+        note_script.clone(),
+        &swap_inputs,
+        &vec![offered_asset],
+        serial_num,
+        sender,
+        tag,
+    )?;
 
-    let merge_script = Hasher::merge(&[serial_num_hash, script_hash]);
+    let p2id_note = create_p2id_note(sender, sender, vec![requested_asset], rng)?;
```

Comment:
@bobbinth, here I want to generate the returning P2ID note. My only concern is that considering that anyone on the network could consume the swap note we won't be able to fill in the `sender` field needed to create a P2ID. 

Am I missing something?

Reply:
Ah! You are correct - the `Note` struct currently requires knowing the sender which we won't have at this point. Let me think about how we can refactor this.